### PR TITLE
rustdoc: render bounds of cross-crate GAT params

### PR DESCRIPTION
**Pull request by fmease**
Follow-up to #102439.
Render the trait bounds of type parameters of cross-crate (generic) associated types.

@rustbot label T-rustdoc A-cross-crate-reexports
r? @GuillaumeGomez

### CODE REVIEW

File: src/librustdoc/clean/mod.rs [hunk 1]
```diff
@@ -1201,21 +1201,19 @@ pub(crate) fn clean_middle_assoc_item<'tcx>(
...
-                //                `clean_ty_generics` to call `explicit_item_bounds` by itself.
                 let bounds = tcx.explicit_item_bounds(assoc_item.def_id);
-                let predicates = ty::GenericPredicates { parent: None, predicates: bounds };
-                let mut generics =
-                    clean_ty_generics(cx, tcx.generics_of(assoc_item.def_id), predicates);
-                // Filter out the bounds that are (likely?) directly attached to the associated type,
-                // as opposed to being located in the where clause.
+                let predicates = tcx.explicit_predicates_of(assoc_item.def_id).predicates;
+                let predicates =
+                    tcx.arena.alloc_from_iter(bounds.into_iter().chain(predicates).copied());
```
Comment (fmease) [inline src/librustdoc/clean/mod.rs:1207]: I have to manually merge the result of `tcx.explicit_item_bounds` and `tcx.explicit_predicates_of` even though there is a query that already does both, namely [`tcx.trait_explicit_predicates_and_bounds`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/query/struct.TyCtxtEnsure.html#method.trait_explicit_predicates_and_bounds). However, the latter only works for *local* definitions (and therefore only accepts `LocalDefId`s) and we obviously need to be able to handle *cross-crate* definitions.

**Comment by fmease** [comment]
Alternatively, I can remove the bound re-sugaring and only keep the `explicit_predicates_of` change. This means that for `type Out14<P: Copy + Eq, Q: ?Sized>;` (from the UI test) we would render the following instead:

```rust
type Out14<P, Q>
where
    P: Copy + Eq,
    Q: ?Sized
```

This would be what we do for (cross-crate) trait parameters. E.g. `trait Tr<T: Eq + Copy> {}` would render a where clause instead of direct bounds.

In any case, as far as I know (and as far as the existing comments vaguely state) there truly is no means to get to know if the user actually wrote the bounds in the where-clause or directly attached to the parameter since this seems to be lost when desugaring, when storing the the crate metadata and/or when the data is *cleaned*. I am not sure which one is the case. Maybe we can reconstruct this by looking at the spans (if available).

Just note that this is a well-known shortcoming of rustdoc's cross-crate documentation abilities. See all <kbd>A-cross-crate-reexports</kbd> issues.

File: src/librustdoc/clean/mod.rs [hunk 2]
```diff
@@ -1273,6 +1271,24 @@ pub(crate) fn clean_middle_assoc_item<'tcx>(
                     }
                     None => bounds.push(GenericBound::maybe_sized(cx)),
                 }
+                // Move bounds that are (likely) directly attached to the parameters of the
+                // (generic) associated type from the where clause to the respective parameter.
+                // There is no guarantee that this is what the user actually wrote but we have
+                // no way of knowing.
+                let mut where_predicates = Vec::new();
+                for mut pred in generics.where_predicates.drain(..) {
```
Comment (GuillaumeGomez) [inline src/librustdoc/clean/mod.rs:1279]: Wouldn't a `into_iter()` achieve the same result?

**Comment by GuillaumeGomez** [comment]
Apart from my question, looks all good to me, thanks!

File: src/librustdoc/clean/mod.rs (hunk 2 above)
Comment (fmease) [reply src/librustdoc/clean/mod.rs:1279]: Ah, you are right! Not sure why I thought I couldn't move out of `generics.where_predicates`. Pushing fix right away…

**Comment by fmease** [comment]
ready

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Thanks!

@bors r+

**Pull request merged**